File: c/mpi.c
#include <stdio.h>
#include <mpi.h>

float equations(int rank, float x, float y) {
    float functions[4];

    functions[0] = -0.2 * y * y * y + 2 * x;
    functions[1] = 0.3 * y;
    functions[2] = 0.01 * y * y - x * x;
    functions[3] = x / y;

  return functions[rank];
}

void runge_kutta(int rank, int n, float x0, float y0, float x, float h, float y[])
{
	float k1, k2, k3, k4;

	y[0] = y0;
	for (int i = 1; i < n; i++)
	{
		k1 = h * equations(rank, x0, y[i - 1]);
		k2 = h * equations(rank, x0 + 0.5 * h, y[i - 1] + 0.5 * k1);
		k3 = h * equations(rank, x0 + 0.5 * h, y[i - 1] + 0.5 * k2);
		k4 = h * equations(rank, x0 + h, y[i - 1] + k3);

		y[i] = y[i - 1] + (1.0 / 6.0) * (k1 + 2 * k2 + 2 * k3 + k4);

		x0 = x0 + h;
	}
}

double get_average(double array[]) {
    double sum = 0.0;
    int size = 10;

    for (int i = 0; i < size; ++i) {
        sum += array[i];
    }

    return sum / size;
}

void test(int rank, int size, int n, float x0, float y0, float x, float h) {
        float y[n];
        double start_time = MPI_Wtime();
        runge_kutta(rank, n, x0, y0, x, h, y);

        float gathered_y[size][n];
        MPI_Gather(y, n, MPI_FLOAT, gathered_y, n, MPI_FLOAT, 0, MPI_COMM_WORLD);

        double end_time;
        if (rank == 0) {
            double end_time = MPI_Wtime();

            printf("x: %f\n", x);
            printf("Time is: %0.8f\n", end_time - start_time);
        }
}

int main(int argc, char *argv[]) {
    int rank, size;

    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    float x0 = 0, x = 100, h = 0.1;
    float y0[] = {0.5, 15, 10, -20};
    int n = (int)((x - x0) / h) + 1;

        float y[n];
        double start_time = MPI_Wtime();
        runge_kutta(rank, n, x0, y0[rank], x, h, y);

        float gathered_y[size][n];
        MPI_Gather(y, n, MPI_FLOAT, gathered_y, n, MPI_FLOAT, 0, MPI_COMM_WORLD);

        double end_time;
        if (rank == 0) {
            double end_time = MPI_Wtime();

            printf("x: %f\n", x);
            printf("Time is: %0.8f\n", end_time - start_time);
                MPI_Finalize();
        }

	return 0;
}

File: python/mpi.py
from mpi4py import MPI
import numpy as np
import time

def equations(x, y):
    functions = [
        -0.2*y*y*y + 2*x,
        0.3*y,
        0.01*y*y - x**2,
        x / y
    ]
    return functions[rank]

def runge_kutta(f, y0, x0, xn, h):
    count = int((xn - x0) / h) + 1
    x_values = [x0]
    y_values = [y0]

    for i in range(1, count):
        x = x_values[i - 1]
        y = y_values[i - 1]

        k1 = h * f(x, y)
        k2 = h * f(x + h/2, y + k1/2)
        k3 = h * f(x + h/2, y + k2/2)
        k4 = h * f(x + h, y + k3)

        y_new = y + (k1 + 2*k2 + 2*k3 + k4)/6
        x_new = x + h

        x_values.append(x_new)
        y_values.append(y_new)

    return np.array(y_values)

def test(xn):
    times = []
    if rank == 0:
        times = []

    for i in range(1, 10):
        start_time = 0
        if rank == 0:
            start_time = time.time()

        y = runge_kutta(equations, y0[rank], x0, xn, h)
        y = comm.gather(y, root=0)

        if rank == 0:
            y = np.vstack(y).T.tolist()

            end_time = time.time()
            times.append(end_time - start_time)

    average = 0
    if rank == 0:
        average = sum(times)/len(times)
    return average

comm = MPI.COMM_WORLD
rank = comm.Get_rank()
size = comm.Get_size()

y0 = [0.5, 15, 10, -20]
x0 = 0
xn = [1000, 5000, 10000, 25000, 50000]
h = 0.1

if rank == 0:
    print('=== Test result ===')
    print('Xn', 'Average time')

for i in range(0, len(xn)):
    average = test(xn[i])
    if rank == 0:
        print(xn[i], average)

if rank == 0:
    print('===================')